{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "431ef080",
   "metadata": {},
   "source": [
    "# Microcompartment Database (BMC-DB)\n",
    "\n",
    "## Project Description\n",
    "\n",
    "This project provides a database framework for storing and linking information\n",
    "about **bacterial microcompartment (BMC) proteins**, their genes, modifications,\n",
    "isoforms, complexes, and external database references.\n",
    "\n",
    "### From microcomp_db.py:\n",
    "\n",
    "This script is the main entry point for the database project. It is responsible\n",
    "for creating the database, reading data from a CSV file, and linking the data to\n",
    "the database. We'll use click to manage the command line interface and SQLAlchemy\n",
    "to manage the database connection and operations.\n",
    "\n",
    "### From db.py:\n",
    "\n",
    "This script is the development of the BMC database using SQLAlchemy.\n",
    "\n",
    "### From file_and_data.py:\n",
    "\n",
    "This script links the addition of data manually to the BMC using csv from readfile.py to database described in db.\n",
    "\n",
    "### From readfile.py:\n",
    "\n",
    "This script allows for the reading of excel file for manual data addition to the database."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7029cebf",
   "metadata": {},
   "source": [
    "## Installation and getting started\n",
    "\n",
    "BMC-database relies on several other programs, packages, and tools for normal running and for development. While many of these dependencies are installed automatically during setup, some may need to be downloaded and installed separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2f496ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clone the repository\n",
    "git clone <https://github.com/monserrj/BMC-database>\n",
    "\n",
    "# Install dependencies\n",
    "pip install -r requirements.txt\n",
    "# Requirements are sqlalchemy, click, pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59d7e88d",
   "metadata": {},
   "source": [
    "Python Version 3.13\n",
    "Sqlite version 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "839fd0dc",
   "metadata": {},
   "source": [
    "## Database Structure - db.py file\n",
    "\n",
    "SQL schema of the database:\n",
    "\n",
    "![BMC Database Schema](PASTE IMAGE LATER)\n",
    "\n",
    "(https://dbdiagram.io/d/DB_STRUCTURE_052025-682494635b2fc4582f9453e4)\n",
    "\n",
    "\n",
    "**Tables**\n",
    "\n",
    "* Protein: stores protein sequence, accession, structure, canonical/isoform\n",
    "\n",
    "* Xdatabase: external database descriptions\n",
    "\n",
    "* Xref: database cross-references\n",
    "\n",
    "* ProteinXref: protein-to-database link\n",
    "\n",
    "* Cds: gene details linked to protein\n",
    "\n",
    "* Origin: original DNA sequence for engineered CDS\n",
    "\n",
    "* CdsXref: CDS-to-database link\n",
    "\n",
    "* Modif: modifications of engineered CDS\n",
    "\n",
    "* CdsModif: CDS-to-modification link\n",
    "\n",
    "* Name: protein names\n",
    "\n",
    "* ProteinName: protein-to-name link\n",
    "\n",
    "* Isoforms: canonical vs isoform proteins\n",
    "\n",
    "* Complex: BMC protein complexes details\n",
    "\n",
    "* ProteinComplex: protein-to-complex link\n",
    "\n",
    "* Interaction: interactions between proteins\n",
    "\n",
    "* Prot_prot_interact: link between protein–protein via interactions\n",
    "\n",
    "**Functions for addition of data to the database**\n",
    "\n",
    "These functions should check that the information added to the database is not duplicate and relationships are correctly added\n",
    "\n",
    "<div class=\"alert alert-block alert-danger\"> \n",
    "<b>Not tested yet\n",
    "\n",
    "* protein_addition(protseq, protaccession, struct, canonical)\n",
    "\n",
    "* xdatabase_addition(xdb_name, href, xdb_type)\n",
    "\n",
    "* cds_addition(cdsseq, cdsaccession, cds_origin, protein)\n",
    "\n",
    "* xref_addition(xdb, cds, protein, xrefacc)\n",
    "\n",
    "* name_addition(protname, protein)\n",
    "</div>\n",
    "\n",
    "**Functions to create database and obtaine the session**\n",
    "* create_db()\n",
    "* get_session()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22a35e3a",
   "metadata": {},
   "source": [
    "## Addition of data manually to the database\n",
    "\n",
    "**Module: file_and_data.py**\n",
    "\n",
    "Links manual data addition to the BMC database using CSV files.\n",
    "\n",
    "*Functions:* link_db_csv() – adds data in a loop from excel file\n",
    "\n",
    "**Module: readfile.py**\n",
    "\n",
    "Reads Excel/CSV files for manual data addition to the database.\n",
    "\n",
    "*Functions:* cli_open_csvfile() – prompts for file path, loads CSV data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ecf110e",
   "metadata": {},
   "source": [
    "## microcomp_db.py\n",
    "\n",
    "Main entry point for project. Handles:\n",
    "\n",
    "* Database creation\n",
    "\n",
    "* Reading CSV input\n",
    "\n",
    "* Linking data to database\n",
    "\n",
    "Uses click for CLI and SQLAlchemy\n",
    "\n",
    "**Running the script**\n",
    "\n",
    "```python microcomp_db.py --dbpath my_database.db --csvpath data/proteins.csv --force --verbose```\n",
    "\n",
    "**Available options**\n",
    "\n",
    "* --dbpath PATH: Path where the SQLite database will be created.\n",
    "\n",
    "    * Default: bmc.db (in current folder).\n",
    "\n",
    "* --csvpath PATH: Path to the CSV file with protein data to be imported.\n",
    "\n",
    "    * Default: ../data/raw/prot_info/prot_data_minimal_correct_UP.csv.\n",
    "\n",
    "* --force / --no-force:\n",
    "\n",
    "    * If --force is set, the script will overwrite an existing database file.\n",
    "    * If --no-force (default), the script will exit without overwriting.\n",
    "\n",
    "* --verbose / --no-verbose:\n",
    "\n",
    "    * If --verbose is set, extra information will be printed during execution.\n",
    "\n",
    "**Workflow**\n",
    "\n",
    "1. Create or overwrite the database (depending on --force).\n",
    "\n",
    "2. Read the CSV file using read_file.\n",
    "\n",
    "3. Link the data into the database using link_db_csv.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "114caeba",
   "metadata": {},
   "source": [
    "## Usage examples\n",
    "\n",
    "Unsure if to do examples by sections \n",
    "\n",
    "### Example 1 - Adding a protein\n",
    "From db.py → protein_addition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "173d4dd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_addition(\n",
    "    protseq=\"MTEYKLVVVGAGGVGKSALTIQLIQNHFVDEYDPT\",\n",
    "    protaccession=\"P12345\",\n",
    "    struct=\"BMC-H\",\n",
    "    canonical=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f592696",
   "metadata": {},
   "source": [
    "### Example 2 – Adding an external database\n",
    "\n",
    "From db.py → xdatabase_addition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b097b97",
   "metadata": {},
   "outputs": [],
   "source": [
    "xdatabase_addition(\n",
    "    xdb_name=\"UniProt\",\n",
    "    href=\"https://www.uniprot.org/\",\n",
    "    xdb_type=\"seq\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de3c4f1b",
   "metadata": {},
   "source": [
    "### Example 3 - Create a new database with default CSV:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78e34f3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "python microcomp_db.py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e4c83fb",
   "metadata": {},
   "source": [
    "### Example 4 - Create a database with custom name and CSV file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38b12769",
   "metadata": {},
   "outputs": [],
   "source": [
    "python microcomp_db.py --dbpath results/bmc_data.db --csvpath data/raw/complex_data.csv\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de9efc37",
   "metadata": {},
   "source": [
    "### Example 5 - Force overwrite an existing database with verbose logging:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "487e04e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "python microcomp_db.py --dbpath bmc.db --csvpath data/raw/complex_data.csv --force --verbose\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90c98c7e",
   "metadata": {},
   "source": [
    "# PENDING SECTIONS:\n",
    "\n",
    "## Dictionaries for when there is vocabulary lock/index\n",
    "This includes convention names, how accession number works, locked vocabulary......\n",
    "## \n",
    "## Contribution Guidelines\n",
    "## Licence\n",
    "## Contact Information\n",
    "## Credits and Acknowledgments\n",
    "* Leighton\n",
    "* SQLAlchemy\n",
    "* Click\n",
    "* Pandas"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
